File: trunk/drools/drools-core/src/main/org/drools/WorkingMemory.java
package org.drools;

/*
 * $Id: WorkingMemory.java,v 1.32 2004-11-08 11:08:29 mproctor Exp $
 *
 * Copyright 2001-2003 (C) The Werken Company. All Rights Reserved.
 *
 * Redistribution and use of this software and associated documentation
 * ("Software"), with or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain copyright statements and
 * notices. Redistributions must also contain a copy of this document.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. The name "drools" must not be used to endorse or promote products derived
 * from this Software without prior written permission of The Werken Company.
 * For written permission, please contact bob@werken.com.
 *
 * 4. Products derived from this Software may not be called "drools" nor may
 * "drools" appear in their names without prior written permission of The Werken
 * Company. "drools" is a trademark of The Werken Company.
 *
 * 5. Due credit should be given to The Werken Company. (http://werken.com/)
 *
 * THIS SOFTWARE IS PROVIDED BY THE WERKEN COMPANY AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE WERKEN COMPANY OR ITS CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

import org.drools.event.WorkingMemoryEventListener;
import org.drools.spi.AgendaFilter;

import java.io.Serializable;
import java.util.List;
import java.util.Map;

/**
 * A knowledge session for a <code>RuleBase</code>.
 *
 * @author <a href="mailto:bob@eng.werken.com">bob mcwhirter </a>
 */
public interface WorkingMemory extends Serializable
{

    /**
     * add event listener to listeners ArrayList
     *
     * @param listener
     */
    public void addEventListener(WorkingMemoryEventListener listener);

    /**
     * remove event listener from listeners ArrayList
     *
     * @param listener
     */
    public void removeEventListener(WorkingMemoryEventListener listener);

    /**
     * Returns a list of listeners
     *
     * @return listeners
     */
    public List getListeners();

    /**
     * Retrieve all of the set application data in this memory
     *
     * @return the application data as a Map
     */
    public Map getApplicationDataMap();

    /**
     * Set a specific piece of application data in this working memory
     *
     * @param name the name under which to populate the data
     * @param value the application data
     */
    public void setApplicationData(String name, Object value);

    /**
     * Retrieve a specific piece of application data by name
     *
     * @return application data or null if nothing is set under this name
     */
    public Object getApplicationData(String name);

    /**
     * Retrieve the <code>RuleBase</code> of this working memory.
     *
     * @return The <code>RuleBase</code>.
     */
    RuleBase getRuleBase();

    /**
     * Fire all items on the agenda until empty.
     *
     * @throws FactException If an error occurs.
     */
    void fireAllRules() throws FactException;

    /**
     * Retrieve the object associated with a <code>FactHandle</code>.
     *
     * @see #containsObject
     *
     * @param handle The fact handle.
     *
     * @return The associated object.
     *
     * @throws NoSuchFactObjectException If no object is known to be associated
     *         with the specified handle.
     */
    Object getObject(FactHandle handle) throws NoSuchFactObjectException;

    /**
     * Retrieve the <code>FactHandle</code> associated with an Object.
     *
     * @see #containsObject
     *
     * @param object The object.
     *
     * @return The associated fact handle.
     *
     * @throws NoSuchFactHandleException If no handle is known to be
     *         associated with the specified object.
     */
    FactHandle getFactHandle(Object object) throws NoSuchFactHandleException;

    /**
     * Retrieve all known objects.
     *
     * @return The list of all known objects.
     */
    List getObjects();

    /**
     * Retrieve all known objects of the specified class.
     *
     * @param objectClass The class of object to return.
     *
     * @return The list of all known objects of the specified class.
     */
    List getObjects(Class objectClass);

    /**
     * Retrieve all known Fact Handles.
     *
     * @return The list of all known fact handles.
     */
    List getFactHandles();

    /**
     * Determine if an object is associated with a <code>FactHandle</code>.
     *
     * @param handle The fact handle.
     *
     * @return <code>true</code> if an object is known to be associated with
     *         the specified handle, otherwise <code>false</code>.
     */
    boolean containsObject(FactHandle handle);

    /**
     * Aasert a fact.
     *
     * @param object The fact object.
     *
     * @return The new fact-handle associated with the object.
     *
     * @throws FactException If an error occurs.
     */
    FactHandle assertObject(Object object) throws FactException;

    /**
     * Retract a fact.
     *
     * @param handle The fact-handle associated with the fact to retract.
     *
     * @throws FactException If an error occurs.
     */
    void retractObject(FactHandle handle) throws FactException;

    /**
     * Modify a fact.
     *
     * @param handle The fact-handle associated with the fact to modify.
     * @param object The new value of the fact.
     *
     * @throws FactException If an error occurs.
     */
    void modifyObject(FactHandle handle, Object object) throws FactException;

    /**
     * Clear the Agenda
     *
     */
    void clearAgenda();
    
    void addAgendaFilter(AgendaFilter agendaFilter);
    
    void removeAgendaFilter(AgendaFilter agendaFilter);
}

File: trunk/drools/drools-dotnet/src/drools.dotnet/WorkingMemory.cs
using System;
using java.util;
using System.Collections.Generic;
using org.drools.dotnet.util;
using org.drools.dotnet.events;

namespace org.drools.dotnet
{
	public class WorkingMemory
	{
		private org.drools.WorkingMemory _javaWorkingMemory;

		/// <summary>
		/// Create a new instance of WorkingMemory
		/// </summary>
		/// <param name="javaWorkingMemory">Java WorkingMemory class</param>
		internal WorkingMemory(org.drools.WorkingMemory javaWorkingMemory)
		{
			_javaWorkingMemory = javaWorkingMemory;
			_javaWorkingMemory.addEventListener(new EventListener(this));
		}

		/// <summary>
		/// Occurs when an object is asserted to working memory
		/// </summary>
		public event EventHandler<ObjectAssertedEventArgs> ObjectAsserted;

		/// <summary>
		/// Occurs when an object is modified in working memory
		/// </summary>
		public event EventHandler<ObjectModifiedEventArgs> ObjectModified;

		/// <summary>
		/// Occurs when an object is retracted from working memory
		/// </summary>
		public event EventHandler<ObjectRetractedEventArgs> ObjectRetracted;

		/// <summary>
		/// Asserts an object into working memory
		/// </summary>
		/// <param name="o">Object to assert</param>
		public void AssertObject(object o)
		{
			_javaWorkingMemory.assertObject(o);
		}


		/// <summary>
		/// Clear the agenda
		/// </summary>
		public void ClearAgenda()
		{
			_javaWorkingMemory.clearAgenda();
		}

		/// <summary>
		/// Determines whether the object is contained in working memory 
		/// </summary>
		/// <param name="o">The object to locate in working memory</param>
		/// <returns>true if the object is found in working memory; otherwise, false. </returns>
		public bool ContainsObject(object o)
		{
			FactHandle factHandle;
			try
			{
				factHandle = _javaWorkingMemory.getFactHandle(o);
				return _javaWorkingMemory.containsObject(factHandle);
			}
			catch (NoSuchFactHandleException)
			{
				return false;
			}
		}

		/// <summary>
		/// Fire all items on the agenda until empty.
		/// </summary>
		public void FireAllRules()
		{
			_javaWorkingMemory.fireAllRules();
		}

		//TODO: fireAllRules(AgendaFilter af)

		/// <summary>
		/// A READ-ONLY copy of all application data
		/// </summary>
		public IDictionary<string, object> ApplicationData
		{
			get{ return new ReadOnlyDictionary<string, object>(
				_javaWorkingMemory.getApplicationDataMap()); }
		}

		/// <summary>
		/// Sets an application data item to a value
		/// </summary>
		/// <param name="identifier">Identifier of the item to set</param>
		/// <param name="value">Value to set the item to</param>
		public void SetApplicationData(string identifier, object value)
		{
			_javaWorkingMemory.setApplicationData(identifier, value);
		}

		/// <summary>
		/// A READ-ONLY list of all objects in working memory.
		/// </summary>
		public IList<object> Objects
		{
			get { return new ReadOnlyList<object>(_javaWorkingMemory.getObjects()); }
		}

		/// <summary>
		/// Retrieves all objects of a specific type from working memory.
		/// </summary>
		/// <typeparam name="T">Type of objects to return.</typeparam>
		/// <returns>List of objects found in working memory of the given type.</returns>
		public IList<T> GetObjects<T>()
		{
			return new ReadOnlyList<T>(_javaWorkingMemory.getObjects(
				java.lang.Class.forName(typeof(T).AssemblyQualifiedName)));
		}

		/// <summary>
		/// RuleBase for the working memory
		/// </summary>
		public RuleBase RuleBase
		{
			get { return new RuleBase(_javaWorkingMemory.getRuleBase()); }
		}

		/// <summary>
		/// Modify an object that already exists in working memory.
		/// </summary>
		/// <param name="o">The modified object</param>
		public void ModifyObject(object o)
		{
			_javaWorkingMemory.modifyObject(_javaWorkingMemory.getFactHandle(o), o);
		}

		/// <summary>
		/// Retract an asserted from from working memory
		/// </summary>
		/// <param name="o">The object to retract</param>
		public void RetractObject(object o)
		{
			_javaWorkingMemory.retractObject(_javaWorkingMemory.getFactHandle(o));
		}

		//TODO: setAsyncExceptionHandler

		# region Event Callouts
		internal void OnObjectAsserted(object sender, ObjectAssertedEventArgs e)
		{
			if (ObjectAsserted != null) ObjectAsserted(sender, e);
		}

		internal void OnObjectModified(object sender, ObjectModifiedEventArgs e)
		{
			if (ObjectModified != null) ObjectModified(sender, e);
		}

		internal void OnObjectRetracted(object sender, ObjectRetractedEventArgs e)
		{
			if (ObjectRetracted != null) ObjectRetracted(sender, e);
		}
		# endregion
	}
}
